python/ray/train/base_trainer.py

```diff
@@ -300,7 +489,10 @@ def preprocess_datasets(self) -> None:
 
         if self.preprocessor:
             train_dataset = self.datasets.get(TRAIN_DATASET_KEY, None)
-            if train_dataset:
+            if train_dataset and self.preprocessor.fit_status in (
+                Preprocessor.FitStatus.NOT_FITTED,
+                Preprocessor.FitStatus.PARTIALLY_FITTED,
+            ):
```

Comment:
for my education, can you share a bit why we have this logic both here and in dataset_spec.py?

Reply:
The `DataParallelTrainer` goes through a separate path for preprocessing datasets. It uses a data parallel ingest spec. Other trainers like xgboost will let the underlying framework handle the data ingest (ex: RayDMatrix params can be passed in to the underlying `xgboost_ray` object.)